File: 2023/day1.py
import time
import re

numbers = {
    "one": 1, 
    "two": 2, 
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7, 
    "eight": 8,
    "nine": 9,
    }

def d1(file) -> int:
    with open(file) as f:
        lines = f.readlines()
        
    s = 0
    for line in lines:
        l = line.strip()
        a = re.search("\d|one|two|three|four|five|six|seven|eight|nine",l)[0]
        b = re.search("\d|eno|owt|eerht|ruof|evif|xis|neves|thgie|enin",l[::-1])[0][::-1]
        a_int = numbers[a] if a in numbers.keys() else int(a)
        b_int = numbers[b] if b in numbers.keys() else int(b)
        s += a_int * 10 + b_int
    return s

day = 1
smallFile = f"inputs/small_{day}.txt"
largeFile = f"inputs/large_{day}.txt"
print(d1(smallFile))
start = time.time()
print(d1(largeFile))
middle = time.time()
#print(d1(file))
end = time.time()

print(f"Part 1 runs in {middle - start}s")
print(f"Part 2 runs in {end - middle}s")

File: 2023/day1.hs
import System.IO ()
import Data.List ( elemIndex, isPrefixOf, isSuffixOf )
import Data.Maybe ( fromJust )

main :: IO ()
main = do    
    file <- getLine
    lines <- readLinesFromFile ("inputs/" ++ file ++ ".txt")
    let parsedLines = map parseLine $ lines
    putStrLn $ "Part 1: " ++ show (sum parsedLines)
    let parsed2Lines = map parseWordLine $ lines
    putStrLn $ "Part 2: " ++ show (sum parsed2Lines)

readLinesFromFile :: FilePath -> IO [String]
readLinesFromFile filePath = do
    contents <- readFile filePath
    return (lines contents)

getNumberFromLine :: String -> Int
getNumberFromLine [] = 0
getNumberFromLine xs = read [head xs]

parseLine :: String -> Int
parseLine line = do
    let numbers = filter (\x -> x >= '0' && x <= '9') line
    let reversed = reverse numbers
    10 * getNumberFromLine numbers + getNumberFromLine reversed

parseWordLine :: String -> Int
parseWordLine line = do
    let digitWords = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
        match = lrcrawl line digitWords
        leftDigit = if length match == 1 then match else show (1 + fromJust (elemIndex match digitWords))
        match2 = rlcrawl line digitWords
        rightDigit = if length match2 == 1 then match2 else show (1 + fromJust (elemIndex match2 digitWords))
    read (leftDigit ++ rightDigit) :: Int

lrcrawl :: String -> [String] -> String
lrcrawl [] words = ""
lrcrawl line words = do
    let matches = filter (\x -> isPrefixOf x line) words
    if length matches > 0 then head matches else lrcrawl (tail line) words

rlcrawl :: String -> [String] -> String
rlcrawl [] words = ""
rlcrawl line words = do
    let matches = filter (\x -> isSuffixOf x line) words
    if length matches > 0 then head matches else rlcrawl (init line) words